src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPrice.cs

```diff
@@ -0,0 +1,176 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Jint.Native;
+using Nethermind.Blockchain.Find;
+using Nethermind.Core;
+using Nethermind.Int256;
+using Nethermind.TxPool;
+
+namespace Nethermind.JsonRpc.Modules.Eth
+{
+    public class GasPrice
+    {
+        private static UInt256? _lastPrice = null; //will this be okay if it is static?
+        private static Block? _lastHeadBlock = null;
+        private readonly IBlockFinder _blockFinder;
+        public GasPrice(IBlockFinder blockFinder)
+        {
+            _blockFinder = blockFinder;
+        }
+        private bool eth_addTxFromBlockToSet(Block block, ref SortedSet<UInt256> sortedSet, UInt256 finalPrice,
+            UInt256? ignoreUnder = null, long? maxCount = null)
+        {
+            ignoreUnder ??= UInt256.Zero;
+            int length = block.Transactions.Length;
+            int added = 0;
+            if (length > 0)
+            {
+                for (int index = 0; index < length && (maxCount == null || sortedSet.Count < maxCount); index++)
+                {
+                    Transaction transaction = block.Transactions[index];
+                    if (!transaction.IsEip1559 && transaction.GasPrice >= ignoreUnder)
+                    {
+                        sortedSet.Add(transaction.GasPrice);
+                        added++;
+                    }
+                }
+
+                switch (added)
+                {
+                    case 0:
+                        sortedSet.Add(finalPrice);
+                        return (false);
+                    case 1:
+                        return (false);
+                    default:
+                        return true;
+                }
+            }
+            else
+            {
+                sortedSet.Add(finalPrice);
+                return false;
+            }
+        }
+
+        private UInt256? eth_latestGasPrice(long headBlockNumber, long genesisBlockNumber)
+        {
+            while (headBlockNumber >= genesisBlockNumber) //should this be latestBlock or headBlock?
```

Comment:
genesis block always has number 0

also a loop like this -> you need to limit this loop since now it will go very far back in time on chains that have many empty blocks -> ideally you would limit the search to 8 / 16 blocks and in such cases if there have been no transactions then you should set the price to minimum